File: app/src/am_sonhk/com/example/asian/adapter/DiffCallback.java
package com.example.asian.adapter;

import androidx.recyclerview.widget.DiffUtil;

import java.util.List;

public class DiffCallback extends DiffUtil.Callback {

    private final List<String> oldList;
    private final List<String> newList;

    public DiffCallback(List<String> oldList, List<String> newList) {
        this.oldList = oldList;
        this.newList = newList;
    }

    @Override
    public int getOldListSize() {
        return oldList.size();
    }

    @Override
    public int getNewListSize() {
        return newList.size();
    }

    @Override
    public boolean areItemsTheSame(int oldItemPosition, int newItemPosition) {
        // Compare items based on their unique identifier
        return oldList.get(oldItemPosition).equals(newList.get(newItemPosition));
    }

    @Override
    public boolean areContentsTheSame(int oldItemPosition, int newItemPosition) {
        // Compare items based on their content
        return oldList.get(oldItemPosition).equals(newList.get(newItemPosition));
    }
}

File: app/src/am_sonhk/com/example/asian/kotlin/diff/DiffCallback.kt
package com.example.asian.kotlin.diff

import androidx.recyclerview.widget.DiffUtil
import com.example.asian.kotlin.model.User

class UserDiffCallback(
    private val oldList: List<User>,
    private val newList: List<User>
) : DiffUtil.Callback() {

    override fun getOldListSize(): Int = oldList.size

    override fun getNewListSize(): Int = newList.size

    override fun areItemsTheSame(oldItemPosition: Int, newItemPosition: Int): Boolean {
        // Assuming User ID is the unique identifier for each user
        return oldList[oldItemPosition].userId == newList[newItemPosition].userId
    }

    override fun areContentsTheSame(oldItemPosition: Int, newItemPosition: Int): Boolean {
        // Check if contents are the same (e.g. name and age are unchanged)
        return oldList[oldItemPosition] == newList[newItemPosition]
    }
}

